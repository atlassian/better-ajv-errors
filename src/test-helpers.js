import { readFileSync } from 'fs';
import { getFixturePath } from 'jest-fixtures';
import Ajv from 'ajv';

export async function getSchemaAndData(name, dirPath) {
  const schemaPath = await getFixturePath(dirPath, name, 'schema.json');
  const schema = JSON.parse(readFileSync(schemaPath, 'utf8'));
  const dataPath = await getFixturePath(dirPath, name, 'data.json');
  const data = JSON.parse(readFileSync(dataPath, 'utf8'));

  return [schema, data];
}

export function evaluateSchema(schema, data) {
  const ajv = new Ajv();
  const fn = ajv.compile(schema);
  fn(data);
  expect(fn.errors.length).toBeGreaterThanOrEqual(0);
  return fn.errors;
}
