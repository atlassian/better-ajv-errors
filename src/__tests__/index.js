import Ajv from 'ajv';
import { getSchemaAndData } from '../test-helpers';
import betterAjvErrors from '../';

describe('Main', () => {
  it('should output error with reconstructed codeframe', async () => {
    const [schema, data] = await getSchemaAndData('default', __dirname);
    const ajv = new Ajv({ jsonPointers: true });
    const validate = ajv.compile(schema);
    const valid = validate(data);
    expect(valid).toBeFalsy();

    const res = betterAjvErrors(schema, data, validate.errors, {
      format: 'cli',
      indent: 2,
    });
    expect(res).toMatchSnapshot();
  });

  it('should output error with codeframe', async () => {
    const [schema, data, json] = await getSchemaAndData('default', __dirname);
    const ajv = new Ajv({ jsonPointers: true });
    const validate = ajv.compile(schema);
    const valid = validate(data);
    expect(valid).toBeFalsy();

    const res = betterAjvErrors(schema, data, validate.errors, {
      format: 'cli',
      json,
    });
    expect(res).toMatchSnapshot();
  });
});
