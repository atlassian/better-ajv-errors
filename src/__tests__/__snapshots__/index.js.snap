// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should output error with codeframe 1`] = `
"[31m[1mENUM[22m[39m[31m should be equal to one of the allowed values[39m
[31m(paragraph, bulletList, mediaSingle, codeBlock, orderedList, heading, panel, blockquote, rule, mediaGroup, applicationCard, decisionList, taskList, table, extension, bodiedExtension, blockCard, layoutSection)[39m

[0m [90m 2 |[39m   [32m\\"type\\"[39m[33m:[39m [32m\\"doc\\"[39m[33m,[39m[0m
[0m [90m 3 |[39m   [32m\\"version\\"[39m[33m:[39m [35m1[39m[33m,[39m[0m
[0m[31m[1m>[22m[39m[90m 4 |[39m   [32m\\"content\\"[39m[33m:[39m [{ [32m\\"type\\"[39m[33m:[39m [32m\\"paragarph\\"[39m }][0m
[0m [90m   |[39m                         [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m [31m[1m👈🏽  Did you mean [95mparagraph[31m here?[22m[39m[0m
[0m [90m 5 |[39m }[0m
[0m [90m 6 |[39m[0m"
`;

exports[`Main should output error with reconstructed codeframe 1`] = `
"[31m[1mENUM[22m[39m[31m should be equal to one of the allowed values[39m
[31m(paragraph, bulletList, mediaSingle, codeBlock, orderedList, heading, panel, blockquote, rule, mediaGroup, applicationCard, decisionList, taskList, table, extension, bodiedExtension, blockCard, layoutSection)[39m

[0m [90m 4 |[39m   [32m\\"content\\"[39m[33m:[39m [[0m
[0m [90m 5 |[39m     {[0m
[0m[31m[1m>[22m[39m[90m 6 |[39m       [32m\\"type\\"[39m[33m:[39m [32m\\"paragarph\\"[39m[0m
[0m [90m   |[39m               [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m [31m[1m👈🏽  Did you mean [95mparagraph[31m here?[22m[39m[0m
[0m [90m 7 |[39m     }[0m
[0m [90m 8 |[39m   ][0m
[0m [90m 9 |[39m }[0m"
`;
